File: Econovation_Algorithm_Mentoring/7. 그래프의 탐색 (DFS, BFS)/Programmers_TargetNumber.java
import java.util.Arrays;

public class Programmers_TargetNumber {
    static char[] arr;
    static int Target;
    static int count;

    public int solution(int[] numbers, int target) {
        int answer = 0;
        arr = new char[]{'+', '-'};
        Target = target;
        count = 0;
        char[] temp = new char[numbers.length];
        makeCombination(target, temp, 0, 0, numbers);
        return count;
    }

    public void makeCombination(int r, char[] temp, int current, int start, int[] numbers) {
        if (temp.length == current) {
            int sum = 0;
            for (int i = 0; i < numbers.length; i++) {
                if (temp[i] == '-') {
                    sum -= numbers[i];
                } else {
                    sum += numbers[i];
                }
            }
            if (Target == sum) {
                count++;
            }
        } else {
            for (int i = 0; i < arr.length; i++) {
                temp[current] = arr[i];
                makeCombination(r, temp, current + 1, i, numbers);
            }
        }
    }
}

File: Algorithm_Swift/Algorithm_Swift/Programmers/Programmers_TargetNumber.swift
//
//  Programmers_TargetNumber.swift
//  Algorithm_Swift
//
//  Created by asong on 2022/03/24.
//

import Foundation

func Programmers_TargetNumber(){
    let arr = ["+","-"]
    var num: [Int] = []
    var t = 0
    var answer = 0
    func solution(_ numbers:[Int], _ target:Int) -> Int {
        num = numbers
        t = target
        let temp = Array(repeating: "0", count: numbers.count)
        makePermutation(0, numbers.count, temp)
        return answer
    }
    func makePermutation(_ count: Int, _ k: Int, _ Temp: [String]){
        var temp = Temp
        if count == k{
            var sum = 0
            for i in 0..<temp.count{
                if temp[i] == "+"{
                    sum += num[i]
                }else{
                    sum -= num[i]
                }
            }
            if sum == t{
                answer += 1
            }
        }else{
            for i in 0..<arr.count{
                temp[count] = arr[i]
                makePermutation(count+1, k, temp)
            }
        }
    }
}
